url9d32402837_dusan_rychnovsky_chess_endgames_tgz-pJ8-cz_dusanrychnovsky_chessendgames_core_strategies_PrecomputedValuesJ8/src/cz/dusanrychnovsky/chessendgames/core/Range.java

```diff
@@ -0,0 +1,156 @@
+package cz.dusanrychnovsky.chessendgames.core;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * 
+ * @author Dušan Rychnovský
+ *
+ */
+public class Range implements Iterable<Position>, Serializable
+{
+	private static final long serialVersionUID = 1L;
+	
+	private final List<Position> positions = new ArrayList<Position>();
+	
+	/**
+	 * Returns a (correctly ordered) range of positions, starting with the
+	 * first and ending with the second given position.
+	 * 
+	 * @param from
+	 * @param to
+	 * @return
+	 */
+	public static Range get(Position from, Position to)
+	{
+		Column fromColumn = from.getColumn();
+		Row fromRow = from.getRow();
+		
+		Column toColumn = to.getColumn();
+		Row toRow = to.getRow();
+		
+		if (fromRow.equals(toRow)) {
+			return buildRowRange(fromColumn, fromRow, toColumn, toRow);
+		}
+		
+		if (fromColumn.equals(toColumn)) {
+			return buildColumnRange(fromColumn, fromRow, toColumn, toRow);
+		}
+		
+		if (fromColumn.distance(toColumn) == fromRow.distance(toRow)) {
+			return buildDiagonalRange(fromColumn, fromRow, toColumn, toRow);
+		}
+		
+		throw new IllegalArgumentException(
+			"Invalid range - from [" + from + "] to [" + to + "]."
+		);
+	}
+	
+	/**
+	 * 
+	 * @param fromColumn
+	 * @param fromRow
+	 * @param toColumn
+	 * @param toRow
+	 * @return
+	 */
+	private static Range buildDiagonalRange(
+		Column fromColumn, Row fromRow, Column toColumn, Row toRow)
+	{
+		Range result = new Range();
+		
+		Iterator<Row> rowIt = fromRow.to(toRow);
+		Iterator<Column> columnIt = fromColumn.to(toColumn);
+		
+		while (rowIt.hasNext())
+		{
+			Row currRow = rowIt.next();
+			Column currColumn = columnIt.next();
+			
+			result.addPosition(Position.get(currColumn, currRow));
+		}
+		
+		return result;
+	}
+	
+	/**
+	 * 
+	 * @param fromColumn
+	 * @param fromRow
+	 * @param toColumn
+	 * @param toRow
+	 * @return
+	 */
+	private static Range buildColumnRange(
+		Column fromColumn, Row fromRow, Column toColumn, Row toRow)
+	{
+		Range result = new Range();
+		
+		Iterator<Row> rowIt = fromRow.to(toRow);
+		while (rowIt.hasNext()) {
+			Row currRow = rowIt.next();
+			result.addPosition(Position.get(fromColumn, currRow));
+		}
+		
+		return result;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 90:98, 112:120.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.